x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/waitforallocations/ScalingInference.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.inference.waitforallocations;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.xpack.core.ml.action.InferModelAction;
+import org.elasticsearch.xpack.core.ml.inference.assignment.AllocationStatus;
+import org.elasticsearch.xpack.core.ml.inference.assignment.TrainedModelAssignment;
+import org.elasticsearch.xpack.core.ml.inference.assignment.TrainedModelAssignmentMetadata;
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+import org.elasticsearch.xpack.ml.inference.assignment.TrainedModelAssignmentService;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.function.BiConsumer;
+import java.util.function.Predicate;
+
+import static org.elasticsearch.core.Strings.format;
+
+public class ScalingInference {
+
+    private static final Logger logger = LogManager.getLogger(ScalingInference.class);
+
+    private final Map<String, LinkedBlockingQueue<WaitingRequest>> queueRequests;
+
+    public record WaitingRequest(
+        InferModelAction.Request request,
+        InferModelAction.Response.Builder responseBuilder,
+        TaskId parentTaskId,
+        ActionListener<InferModelAction.Response> listener
+    ) {
+        public String deploymentId() {
+            return request.getId();
+        }
+    }
+
+    private final TrainedModelAssignmentService assignmentService;
+    private final BiConsumer<WaitingRequest, TrainedModelAssignment> queuedConsumer;
+
+    public ScalingInference(
+        TrainedModelAssignmentService assignmentService,
+        BiConsumer<WaitingRequest, TrainedModelAssignment> queuedConsumer
+    ) {
+        this.assignmentService = assignmentService;
+        this.queuedConsumer = queuedConsumer;
+        this.queueRequests = new ConcurrentHashMap<>();
+    }
+
+    public synchronized void waitForAssignment(WaitingRequest request) {
+        var p = queueRequests.computeIfAbsent(request.deploymentId(), k -> new LinkedBlockingQueue<>());
+
+        if (p.isEmpty()) {
+            logger.info("waitForAssignment will wait for condition");
+            p.offer(request);
+            assignmentService.waitForAssignmentCondition(
+                request.deploymentId(),
+                new DeploymentHasAtLeastOneAllocation(request.deploymentId()),
+                request.request().getInferenceTimeout(),
+                new WaitingListener(request.deploymentId())
```

Comment:
I don't think this logic works out perfect. 

What if num allocations=0 and:
- t=0: inference request 1 -> queue, wait for condition
- t=9: inference request 2 -> queue
- t=10: condition failed, both requests fail
- t=11: allocation started

Now the second request has only waited for 1 sec before failing, right? While it could've been handled successfully after 2secs.